io/shared/src/main/scala/fs2/io/process/ProcessBuilder.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2013 Functional Streams for Scala
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2.io
+package process
+
+import cats.effect.kernel.Resource
+import fs2.io.file.Path
+
+sealed abstract class ProcessBuilder private {
+  def command: String
+  def args: List[String]
+  def env: Map[String, String]
+  def workingDirectory: Option[Path]
+
+  def withCommand(command: String): ProcessBuilder
+  def withArgs(args: List[String]): ProcessBuilder
+  def withEnv(env: Map[String, String]): ProcessBuilder
+  def withWorkingDirectory(workingDirectory: Path): ProcessBuilder
+  def withCurrentWorkingDirectory: ProcessBuilder
+
+  final def spawn[F[_]: ProcessSpawn]: Resource[F, Process[F]] =
+    ProcessSpawn[F].spawn(this)
+}
+
+object ProcessBuilder {
+
+  def apply(command: String, args: List[String]): ProcessBuilder =
```

Comment:
Out of curiosity : why not `def apply(command: String, args: String*)` ? Is there some bin-compat related reason that make it bad ?

Reply:
No reason, added 👍

Reply:
I would probably use this api in scripts most often, and I'd like a more high-level api that abstracts some common patterns, e.g. passing optional arguments, or arguments k-v pairs, or options of those, etc.
Unfortunately the way you do that is with a bunch of implicit conversions 😬

Reply:
> I would probably use this api 

I'm not sure if this is the API you want :) It's kind of like the JDK `Process` stuff vs Scalalib process DSL. The goal here is to provide the API and do the cross-platform stuff.

Another lib makes sense for your high-level API. If it's targeting scripts it can break bincompat more freely than a library can.

Reply:
> Unfortunately the way you do that is with a bunch of implicit conversions 

I think this would be a nice client-side addition to decline, or some higher-level library of the sort. To elaborate, Smithy4s provides an integration with decline OOTB  : https://disneystreaming.github.io/smithy4s/docs/protocols/deriving-cli

With this PR being merged to fs2, I'd be inclined to add the client-side of things to go from high-level method calls to low level args that could be used in the ProcessBuilder, in a way that would match what is done in the decline integration. 

Not trying to advertise, just saying that seems like a job for higher-level libraries. In particular because there is no actual "standard" for things like arguments key-value pairs, etc .

Reply:
> I'd be inclined to add the client-side of things to go from high-level method calls to low level args that could be used in the ProcessBuilder, in a way that would match what is done in the decline integration.

Wow, I hadn't even thought about this! 🤩

Reply:
> Another lib makes sense for your high-level API.

Well, sort of.
I don't strongly disagree, but a couple of counterpoints:
- fs2-io started like this, then over time it became more and more a user-facing lib, I don't really see why this would be that much different
- Any additional lib is a pain for scripts, i.e. you have to remember/google id and version before you can hit the ground running.

Reply:
> Any additional lib is a pain for scripts

```
//> using toolkit "typelevel"
```

🙂

Reply:
> Any additional lib is a pain for scripts, i.e. you have to remember/google id and version before you can hit the ground running.

@SystemFw can I interest you in a modest tool I've written for that very reason ? IIRC you're a emacs user so here's the CLI version : https://github.com/neandertech/dexsearch

Reply:
> `//> using toolkit "typelevel"`

Is this a thing?!

Reply:
It will be soon, hopefully :)

Reply:
ICYMI: https://github.com/typelevel/toolkit